R/control_chart.R

```diff
@@ -0,0 +1,139 @@
+#' Defaults to mean
+#' @param sigmas Number of standard deviations above and below the central 
+#' tendency to call a point influenced by "special cause variation."
+#' Defaults to 3
+#' @param save_to Optional file path to save chart. If not provided, the chart
+#' will be printed on screen
+#' @param plot_width In inches. Only used if save_to is specified
+#' @param plot_height In inches. Only used if save_to is specified
+#' @param plot_title Title in upper-left
+#' @param plot_catpion Caption in lower-right
+#' @param plot_font_size Base font size; text elements will be scaled to this
+#'
+#' @return Generally called for the side effect of printing the control chart
+#' or writing the control chart to file, depending on whether save_to is 
+#' specified. Invisibly, returns a ggplot object for further customization.
+#' @export
+#' @import ggplot2
+#' @importFrom readr read_csv
+#' @importFrom tibble data_frame
+#' @importFrom stats as.formula
+#'
+#' @examples
+#' d <- 
+#'   tibble::data_frame(
+#'     day = sample(c("Monday", "Tuesday", "Wednesday", "Thursday", "Friday"), 
+#'                  100, TRUE),
+#'     person = sample(c("Tom", "Jane", "Alex"), 100, TRUE),
+#'     count = rbinom(100, 20, ifelse(day == "Friday", .5, .2)),
+#'     date = Sys.Date() - sample.int(100))
+#' control_chart(d, "count")
+#' control_chart(d, "count", group1 = "day", group2 = "person")
+#' if (require(ggplot2)) {
+#' x <- control_chart(d, "count", "date")
+#' x + 
+#'   ylab("Number of Adverse Events") + 
+#'   scale_x_date(name = "Week of", date_breaks = "week") +
+#'   theme(axis.text.x = 
+#'                    element_text(angle = -90, vjust = 0.5, hjust=1))
+#' }
+control_chart <- function(d, measure, x, group1, group2,
+                          center_line = mean, sigmas = 3,
+                          save_to, plot_width = 8, plot_height = 4,
+                          plot_title = NULL, plot_catpion = NULL,
+                          plot_font_size = 11) {
+
+  theme_set(theme_bw())
+
+  if (missing(d) || !(is.data.frame(d) || is.character(d))) {
+    stop("You have to provide a data frame or a file location.")
+  } else if (is.character(d)) {
+    message("Attempting to read csv from ", d)
+    d <- readr::read_csv(d)
+  }
+
+  if (missing(measure)) {
+    stop("You have to provide a measure variable name.")
+  }
+
+  if (missing(x)) {
+    x <- "x"
+    d$x <- seq_len(nrow(d))
+  }
+
+  bounds <- calculate_bounds(d, measure, center_line, sigmas)
+
+  # Calculate central tendency and upper and lower limits
+  d$outside <- ifelse(d[[measure]] > bounds[["upper"]] |
+                        d[[measure]] < bounds[["lower"]], "out", "in")
+
+  # Make plot
+  chart <-
+    ggplot(d, aes_string(x = x, y = measure)) +
+    geom_hline(yintercept = bounds[["mid"]], color = "darkgray") +
+    geom_hline(yintercept = c(bounds[["upper"]], bounds[["lower"]]),
+                        linetype = "dotted", color = "darkgray") +
+    geom_line() +
```

Comment:
I wonder if we should turn off the line if the data gets larger than a certain size?
![image](https://user-images.githubusercontent.com/15145347/32338542-b65fd934-bfba-11e7-88d5-6548cc678f27.png)

Reply:
I love the plot handling. Flexible, pretty good defaults. I wonder if we should also shift label text to vertical if enough are on there.
![mikesfile](https://user-images.githubusercontent.com/15145347/32338895-ac0a31a4-bfbb-11e7-9522-8e49f60d130b.png)

Reply:
@mmastand let's think more generally about this. The function has arguments to use a bigger canvas, which can handle the overplotting lines problem within reason, and the last example shows how to deal with the over-written x-axis labels by further customizing the output ggplot object. Trying to customize plots for things like long labels and lots of data is very hard. I'm inclined to provide defaults that will work well in typical cases, have arguments for easy customizations, and output an object that can be further customized for all the hundreds of possible edge cases.